src/puppetlabs/puppetdb/query/population.clj

```diff
@@ -11,61 +11,54 @@
             [honeysql.core :as hcore]
             [honeysql.helpers :as hh]))
 
-;;; Query library
-
-(def ^:private from-all-resources-and-their-nodes
-  (hcore/build :from [:certnames]
-               :join [:catalog_resources [:= :certnames.id :catalog_resources.certname_id]]))
-
-(defn- where-nodes-are-active [q]
-  (hh/merge-where q [:and
-                     [:= :certnames.deactivated nil]
-                     [:= :certnames.expired nil]]))
-
-(defn- select-count [q]
-  (-> q
-      (hh/select [:%count.* :c])
-      hcore/format
+(defn first-query-result
+  "Pick a key out of the first result of a query."
+  [query k]
+  (-> query
       query-to-vec
       first
-      :c))
-
-;;; Public
+      k))
 
 (defn num-resources
   "The number of resources in the population"
   []
   {:post [(number? %)]}
-  (-> from-all-resources-and-their-nodes
-      where-nodes-are-active
-      select-count))
+  (-> "select reltuples::bigint as c from pg_class where relname='catalog_resources'"
+      (first-query-result :c)))
 
-(defn num-nodes
+(defn num-inactive-nodes
+  "The number of expired/deactivated nodes"
+  []
+  {:post [(number? %)]}
+  (-> "select count(*) as c from certnames where deactivated is not null or expired is not null"
+      (first-query-result :c)))
+
+(defn num-active-nodes
   "The number of unique certnames in the population"
   []
   {:post [(number? %)]}
-  (-> (hh/from :certnames)
-      where-nodes-are-active
-      select-count))
+  (-> "select count(*) as c from certnames where deactivated is null and expired is null"
+      (first-query-result :c)))
 
 (defn avg-resource-per-node
   "The average number of resources per node"
   []
   {:post [(number? %)]}
-  (quotient (num-resources) (num-nodes)))
+  (let [nresources (num-resources)
+        nnodes (first-query-result "select count(*) as c from certnames" :c)]
+    (/ nresources nnodes)))
 
 (defn pct-resource-duplication
   "What percentage of resources in the population are duplicates"
   []
   {:post [(number? %)]}
-  (let [distinct-resources-q (-> from-all-resources-and-their-nodes
-                                 where-nodes-are-active
-                                 (hh/select :catalog_resources.resource)
-                                 (hh/modifiers :distinct))
-        num-unique (-> (hh/from [distinct-resources-q :r])
-                       select-count)
-        num-total (num-resources)]
-    (quotient (- num-total num-unique) num-total)))
+  (let [nresources (num-resources)
+        ndistinct (-> "select count(*) as c from
+                       (select distinct resource from catalog_resources) dist"
```

Comment:
We were seeing 5-10s queries on the big db, right? If that's correct, I'd jack it up to about a minute.